aiida/engine/daemon/execmanager.py

```diff
@@ -351,7 +354,7 @@ def upload_calculation(
         remotedata.store()
 
 
-def submit_calculation(calculation: CalcJobNode, transport: Transport) -> str:
+def submit_calculation(calculation: CalcJobNode, transport: Transport) -> str | ExitCode:
```

Comment:
About using the exit code pattern here: an alternative could be to just raise an exception of a specific type. AiiDA would then handle that exception and set the exit code of the process (which would be fixed for all such errors) + the message from the exception.

In this approach, the function would not need to return different types of objects, and you may be able to avoid some if/else checks

Reply:
Yes, I agree that would be a valid alternative. I chose the current approach for consistency. The same mechanism is already used for the `Scheduler.parse_output` method. And the code would not necessarily be cleaner or more succinct with the exception approach. The stack going from `Scheduler._parse_submit_output` all the way to where it is handled in `Waiting.execute` is quite deep and there will be quite a bit of exception handling if that is the chosen mechanism.